src/V3DfgSynthesize.cpp

```diff
@@ -374,28 +381,49 @@ class AstToDfgConverter final : public VNVisitor {
         UASSERT_OBJ(m_converting, nodep, "AstToDfg visit called without m_converting");
         UASSERT_OBJ(!nodep->user2p(), nodep, "Already has Dfg vertex");
         if (unhandled(nodep)) return;
-        DfgVertex* const vtxp = make<DfgConst>(nodep->fileline(), nodep->num());
-        nodep->user2p(vtxp);
+
+        if (nodep->width() != nodep->num().width()) {
+            // Sometimes the width of the AstConst is not the same as the
+            // V3Number it holds. Truncate it here. TODO: should this be allowed?
+            V3Number num{nodep, nodep->width()};
+            num.opSel(nodep->num(), nodep->width() - 1, 0);
+            DfgVertex* const vtxp = make<DfgConst>(nodep->fileline(), num);
+            nodep->user2p(vtxp);
```

Comment:
@wsnyder the new type checker highlighted I needed to fix this up here. It seems like a bug but wanted to check with you.

For an AstConst, `nodep->width()` sometimes differs from `nodep->num().width()`. AFAIK V3Const inserts the mismatching ones. If we should indeed be able to assert here that these widths match I'm happy to go back and fix them up in V3Const.

Reply:
Cleanup welcome. Post V3Clean(-ish) the width will be larger than the verilog width which is represented in widthMin, so some V3Number algorithms do need to know the verilog width instead.

Reply:
So:
- Pre ~V3Clean `AstConst::width()` should equal `AstConst::num().width()`
- ~V3Clean onwards `AstConst::width()` can be larger than `AstConst::num().width()`, and even then `AstConst::widthMin()` should equal `AstConst::num().width()`

Is this accurate? (at least as far as we can reason right now)

What I'm seeing is actually the V3Number being wider than the AstConst, which is the opposite.

Reply:
I think it's still true that Const's width() should always match V3Number. Pointing out that some of the V3Number operations need to use minWidth, and it's likely if you clean up you'll find that there is some inconsistencies requiring minWidth to get passed to new V3Numbers and/or places minWidth wasn't used but should have been.